neurodamus/node.py

```diff
@@ -1299,6 +1300,36 @@ def _sim_corenrn_write_config(self, corenrn_restore=False):
                 self._pc.nrnbbcore_write(corenrn_data)
                 MPI.barrier()  # wait for all ranks to finish corenrn data generation
 
+        if self._shm_enabled and SimConfig.cli_options.enable_shm == "CACHE":
+            # Below improvement should only be enabled if /dev/shm is available
+            # If /dev/shm is enabled this means the the <*_{1,2,3}.dat> files are going to
+            # be written first in /dev/shm (which is going to be used as a buffer to
+            # accelerate the writes and avoid GPFS IO)
+            # Rank 0 should move the <*_{1,2,3}.dat> files to a the coreneuron_datadir on
+            # GPFS (coreneuron_input_gpfs). Each node should move them to a separate
+            # subfolder (with name NODE_ID?) of coreneuron_datadir to increase GPFS
+            # performance. We can do that by finding out the Rank 0 of every node (already
+            # implemented).
+            # We will then create links on coreneuron_datadir for each file in the
+            # subfolders.
+            local_node_rank0 = int(SHMUtil.local_ranks[0])
+            if MPI.rank == local_node_rank0:
+                import shutil
+
+                group_id = int(SHMUtil.node_id / 20)
+                node_specific_corenrn_output_in_storage = \
+                    Path(corenrn_output) / f"coreneuron_input/cycle_{self._cycle_i}/group_{group_id}/node_{SHMUtil.node_id}"
```

Comment:
**(Pasting from Slack)**

Thanks to a reminder by @1uc, I completely forgot that I discussed with 
@iomaganaris about generating the output in subfolders on GPFS, and that he already had something working recently 😅. Hence, I made some changes to improve it a bit by dividing the `coreneuron_input` directory into:
```
coreneuron_input/cycle_X/group_Y/node_Z 
```
Being `cycle_X` the current cycle in the model instantiation, `node_Z` the node ID (i.e., from 0 to 799 in the 800-node simulation), and `group_Y` a simple division to group the set of nodes in subfolders dividided by 20 (i.e., `group_id=floor(node_id / 20)`). Why 20? Well, a magic number just to divide the amount of folders into something reasonable.

With this simple approach, inside the `coreneuron_input` there would be **at most** 32 folders (i.e., one per cycle). Inside each cycle folder, there would be **at most** 40 subfolders (i.e., 800/20 = 40, one per subset of nodes). Inside the group of nodes, there would be **at most** 20 subfolders corresponding to the node IDs (i.e., from 0 to 19 in the first group, 20 to 39 in the second, and so on). Finally, inside the specific node folder, there would be **at most** 120 files from CoreNEURON (i.e., 3 files_per_rank x 40 ranks_per_node = 120).

In other words, we go from a folder with 3.1M files, to a tree that is much more manageable by GPFS and IME, regardless of the target file system that we use.

Reply:
Note that there has been a slight update in the code since this comment was written, but the reasoning is still valid.